org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/vue/autoinsert/VueAutoInsertReconciler.java

```diff
@@ -0,0 +1,209 @@
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.TextDocumentIdentifier;
+import org.eclipse.swt.widgets.Display;
+import org.eclipse.wildwebdeveloper.Activator;
+import org.eclipse.wildwebdeveloper.vue.VueLanguageServerAPI;
+
+/**
+ * {@link IReconciler} implementation used to support auto close tags / auto
+ * insert quote, features provides by the vscode VUE language server with the
+ * custom 'volar/client/autoInsert' LSP request.
+ *
+ */
+public class VueAutoInsertReconciler implements IReconciler {
+
+	private IDocument document;
+
+	private ITextViewer viewer;
+
+	private Listener listener;
+
+	private void autoInsert(DocumentEvent event) {
+		if (event == null || viewer == null) {
+			return;
+		}
+		IDocument document = event.getDocument();
+		if (document == null || event == null || event.getLength() != 0) {
+			return;
+		}
+
+		int offset = event.getOffset();
+		URI uri = LSPEclipseUtils.toUri(document);
+		if (uri == null) {
+			return;
+		}
+		
+		
+		TextDocumentIdentifier identifier = new TextDocumentIdentifier(uri.toString());
+		
+		LanguageServers.forDocument(document).collectAll((w, ls) -> CompletableFuture.completedFuture(ls))
+				.thenAccept(lss -> lss.stream().filter(VueLanguageServerAPI.class::isInstance)
+						.map(VueLanguageServerAPI.class::cast).findAny().ifPresent(info -> {
+							// The document is bound with HTML language server, consumes the html/autoInsert
+							final Display display = viewer.getTextWidget().getDisplay();
+							CompletableFuture.supplyAsync(() -> {
+								try {
+									// Wait for textDocument/didChange
+									Thread.sleep(100);
+								} catch (InterruptedException ex) {
+									Thread.interrupted();
+								}
+								try {
+									
+									AutoInsertParams params = new AutoInsertParams();
+									params.setTextDocument(identifier);
+									params.setPosition(LSPEclipseUtils.toPosition(offset + event.getText().length(), document));
+									
+									AutoInsertOptions opts = new AutoInsertOptions();
+									AutoInsertLastChange changeEvent = new AutoInsertLastChange();
+									final var range = new Range(LSPEclipseUtils.toPosition(offset, document),
+											LSPEclipseUtils.toPosition(offset + event.fLength, document));
+									changeEvent.setRange(range);
+									changeEvent.setText(event.getText());
+									changeEvent.setRangeLength(event.fLength);
+									changeEvent.setRangeOffset(offset);
+									opts.setLastChange(changeEvent);
+									params.setOptions(opts);
+
+									// consumes html/autoInsert from HTML language server
+									info.autoInsert(params)
+											.thenAccept(response -> {
+												if (response != null) { 
+													display.asyncExec(() -> {
+														try {
+															// we receive a text like
+															// $0</foo>
+															// $0 should be used for set the cursor.
+															String newText = response.map(l -> l, r -> r.getNewText());

```

Comment:
Can be
```suggestion
															String newText = response.map(Function.identity(), r::getNewText);
```

Reply:
Done, as I wrote in description this patch is much bigger than AutoInsert. Take a look to VuePreferences, and how it work with vscode-vue package.json

Reply:
> Done, as I wrote in description this patch is much bigger than AutoInsert. 

Do you think it's worth reworking this patch in a series of easier to understand yet functionally valuable patches to facilitate further maintenance?

Reply:
Sure, one moment

Reply:
I published autoinsert-only part (works on default settings). For example auto close html tags `<tag>|</tag>`, and transform `{{|}}` to `{{ | }}`  in templates